src/reducers/computeRenderedRows.js

```diff
@@ -207,35 +211,68 @@ function computeRenderedRowOffsets(state, rowRange) {
     return;
   }
 
-  const bufferMapping = []; // state.rows
+  // output for this function
+  const bufferMapping = state.rows;
   const rowOffsetsCache = {}; // state.rowHeights
+
+  // incremental way for calculating rowOffset
   let runningOffset = rowOffsets.sumUntil(firstBufferIdx);
-  for (let rowIdx = firstBufferIdx; rowIdx < endBufferIdx; rowIdx++) {
 
+  for (let rowIdx = firstBufferIdx; rowIdx < endBufferIdx; rowIdx++) {
     // Update the offset for rendering the row
     rowOffsetsCache[rowIdx] = runningOffset;
     runningOffset += storedHeights[rowIdx];
 
-    // Check if row already has a position in the buffer
-    let rowPosition = bufferSet.getValuePosition(rowIdx);
-
-    // Request a position in the buffer through eviction of another row
-    if (rowPosition === null && bufferSet.getSize() >= renderedRowsCount) {
-      rowPosition = bufferSet.replaceFurthestValuePosition(
-        firstViewportIdx,
-        endViewportIdx - 1,
-        rowIdx
-      );
-    }
-
-    // If we can't reuse any existing position, create a new one
-    if (rowPosition === null) {
-      rowPosition = bufferSet.getNewPositionForValue(rowIdx);
+    // we'll skip getting a row position for the rows outside the viewport
+    if (computeWithinViewportOnly && (rowIdx < firstViewportIdx || rowIdx >= endViewportIdx)) {
+      continue;
     }
 
+    // Get position for the viewport row
+    const rowPosition = addRowToBuffer(rowIdx, bufferSet, firstViewportIdx, endViewportIdx, renderedRowsCount);
     bufferMapping[rowPosition] = rowIdx;
   }
 
+  if (computeWithinViewportOnly) {
+    // In the above loop, we only calculated row offsets for the rows inside the buffer.
+    // Now we calculate the row offsets for the remaining rows.
+    const rowsNotInBuffer = filter(state.rows, (rowIdx) => !inRange(rowIdx, firstBufferIdx, endBufferIdx));
```

Comment:
Should this be over bufferMapping instead of state.rows?  Does state.rows risk being out of date at this point?

Reply:
> Should this be over bufferMapping instead of state.rows? 

I think bufferMapping and state.rows will be the same. So there's no point in assigning at the end.

However, to fix the controlled scrolling issue, we throw away the buffer (previous behavior) and use a fresh array. So this means I'll still need the assign at the end.

> Does state.rows risk being out of date at this point?

I don't think that'll happen since the reducer updates are guaranteed to be non-preemptive, if that's what you mean.
Also bufferMapping is always the same as state.rows (and on the new commit, only when computeWithinViewportOnly is true), hence this shouldn't be a problem. 
I'll still update it to bufferMapping as it leads to confusion (I just got confused while looking at this again)

Reply:
I think this gets back to code operating unclearly over rows in viewport vs outside of viewport.  Let's split things out so there's less room for error.  

My concern is you're assuming no code above modifies any entries which are not inRange, but that may not be true for future changes to the repo.  It's better if you write code which makes less assumptions of that nature, or if you can't you should ensure those assumptions are well documented.